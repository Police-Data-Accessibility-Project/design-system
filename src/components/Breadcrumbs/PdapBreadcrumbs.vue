<template>
	<nav aria-label="Breadcrumb">
		<transition-group class="pdap-breadcrumbs" name="pdap-breadcrumbs" tag="ul">
			<li
				v-for="breadcrumb in breadcrumbs"
				:key="breadcrumb.text"
				:class="{ 'is-active': breadcrumb.active }"
			>
				<router-link v-if="!breadcrumb.active" :to="breadcrumb.path">
					{{ breadcrumb.text }}
				</router-link>
				<span v-else>{{ breadcrumb.text }}</span>
			</li>
		</transition-group>
	</nav>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import { useRoute } from 'vue-router';
import { getBreadcrumbs } from '../../utils/breadcrumbs';

const route = useRoute();

const breadcrumbs = computed(() => getBreadcrumbs(route));
</script>

<style scoped>
.pdap-breadcrumbs {
	@apply flex items-center;
}

.pdap-breadcrumbs li {
	@apply flex items-center;
}

.pdap-breadcrumbs li:not(:first-child)::before {
	@apply mx-2;

	content: '/';
}

.pdap-breadcrumbs .is-active {
	@apply text-neutral-950;
}

/* Animations */
.pdap-breadcrumbs-enter-active,
.pdap-breadcrumbs-leave-active {
	transition: opacity 0.2s ease;
}

.pdap-breadcrumbs-enter-from,
.pdap-breadcrumbs-leave-to {
	opacity: 0;
}
</style>
