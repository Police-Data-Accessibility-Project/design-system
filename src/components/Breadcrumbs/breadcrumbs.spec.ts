import { RouterLinkStub, flushPromises, mount } from '@vue/test-utils';
import { createRouter, createWebHistory } from 'vue-router';
import PdapBreadcrumbs from './PdapBreadcrumbs.vue';
import { describe, it, expect } from 'vitest';

const routes = [
	{
		path: '/product',
		name: 'Product',
		meta: { breadcrumbText: 'Product' },
		redirect: undefined,
		children: [
			{
				path: '/product/:id',
				name: 'Product ID',
				meta: { breadcrumbText: 'Product ID' },
				redirect: undefined,
				children: [
					{
						path: '/product/:id/edit',
						name: 'Edit Product ID',
						meta: { breadcrumbText: 'Edit Product ID' },
						redirect: undefined,
						children: [],
					},
				],
			},
		],
	},
];

const router = createRouter({
	history: createWebHistory(),
	routes,
});

describe('PdapBreadcrumbs', () => {
	it('renders breadcrumbs correctly for a root route', async () => {
		router.push({ name: 'Product' });
		await router.isReady();

		const wrapper = mount(PdapBreadcrumbs, {
			global: {
				mocks: {
					$route: routes[0],
				},
				plugins: [router],
				stubs: {
					RouterLink: RouterLinkStub,
				},
			},
		});

		const breadcrumbItems = wrapper.findAll('li');
		expect(breadcrumbItems).toHaveLength(1);
		// @ts-expect-error
		expect(breadcrumbItems.at(0).text()).toBe('Product');
		// @ts-expect-error
		expect(breadcrumbItems.at(0).classes()).toContain('is-active');
	});

	it('renders breadcrumbs correctly for a single-nested route', async () => {
		router.push({ name: 'Product ID', params: { id: 1234 } });
		await router.isReady();
		await flushPromises();

		const wrapper = mount(PdapBreadcrumbs, {
			global: {
				mocks: {
					$route: routes[0].children[0],
				},
				plugins: [router],
				stubs: {
					RouterLink: RouterLinkStub,
				},
			},
		});

		const breadcrumbItems = wrapper.findAll('li');
		expect(breadcrumbItems).toHaveLength(2);
		// @ts-expect-error
		expect(breadcrumbItems.at(0).text()).toBe('Product');
		// @ts-expect-error
		expect(breadcrumbItems.at(1).text()).toBe('Product ID');
		// @ts-expect-error
		expect(breadcrumbItems.at(1).classes()).toContain('is-active');

		expect(wrapper.html()).toMatchSnapshot();
	});

	it('renders breadcrumbs correctly for a double-nested route', async () => {
		router.push({ name: 'Edit Product ID', params: { id: 1234 } });
		await router.isReady();
		await flushPromises();

		const wrapper = mount(PdapBreadcrumbs, {
			global: {
				mocks: {
					$route: routes[0].children[0].children[0],
				},
				plugins: [router],
				stubs: {
					RouterLink: RouterLinkStub,
				},
			},
		});

		const breadcrumbItems = wrapper.findAll('li');
		expect(breadcrumbItems).toHaveLength(3);
		// @ts-expect-error
		expect(breadcrumbItems.at(0).text()).toBe('Product');
		// @ts-expect-error
		expect(breadcrumbItems.at(1).text()).toBe('Product ID');
		// @ts-expect-error
		expect(breadcrumbItems.at(2).text()).toBe('Edit Product ID');
		// @ts-expect-error
		expect(breadcrumbItems.at(2).classes()).toContain('is-active');

		expect(wrapper.html()).toMatchSnapshot();
	});
});
