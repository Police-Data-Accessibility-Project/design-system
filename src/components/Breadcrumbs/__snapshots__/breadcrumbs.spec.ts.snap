// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PdapBreadcrumbs > renders breadcrumbs correctly for a double-nested route 1`] = `
<nav aria-label="Breadcrumb">
  <transition-group-stub appear="false" class="pdap-breadcrumbs" css="true" name="pdap-breadcrumbs" persisted="false" tag="ul">
    <li class>
      <a>Product</a>
    </li>
    <li class>
      <a>Product ID</a>
    </li>
    <li class="is-active">
      <span>Edit Product ID</span>
    </li>
  </transition-group-stub>
</nav>
`;

exports[`PdapBreadcrumbs > renders breadcrumbs correctly for a single-nested route 1`] = `
<nav aria-label="Breadcrumb">
  <transition-group-stub appear="false" class="pdap-breadcrumbs" css="true" name="pdap-breadcrumbs" persisted="false" tag="ul">
    <li class>
      <a>Product</a>
    </li>
    <li class="is-active">
      <span>Product ID</span>
    </li>
  </transition-group-stub>
</nav>
`;
